### DISCUSSION TITLE
Enable Logs for Azure SSO

### DISCUSSION THREAD

**Discussion by sabinayakc** [nautobot/nautobot#6620] [Q&A]
Hello, 

I am trying to integrate Azure AD. So far everything works except group assignments.

I followed the instructions here https://github.com/nautobot/nautobot/blob/main/examples/azure_ad/README.md

How do I enable logging for SSO troubleshooting?

**Comment by jdrew82** [comment]
@sabinayakc I'd suggest modifying your group_sync function to print out the information you need for troubleshooting. That's always worked for me and you can see the exact response you get from your IDP once authenticated. Can you please clarify what issues you're having with group assignments though? It's fairly basic and easy to setup. Based upon the groups that the user is assigned you can define the groups that you wish to use to assign as admin or superuser roles with the STAFF_GROUPS and SUPERUSER_GROUPS. There's also the option for logging as described [here](https://github.com/nautobot/nautobot/blob/main/examples/azure_ad/README.md#displaying-logging-from-my_custom_module) in the README you linked.

**Reply by sabinayakc** [reply]
For some context, I am using helm chart deployment in a K8s cluster. With the following env variables.

```
    - name: "NAUTOBOT_SSO_CLAIMS_GROUP"
      value: "groups"
    - name: "NAUTOBOT_SSO_ENABLE_GROUP_SYNC"
      value: "true"
    - name: "NAUTOBOT_SSO_SUPERUSER_GROUPS"
      value: "Group Name in Azure"
    - name: "NAUTOBOT_SSO_STAFF_GROUPS"
      value: "Group Name in Azure"
 ```
 
 And config
 
 ```
 AUTHENTICATION_BACKENDS = [
"social_core.backends.azuread_tenant.AzureADTenantOAuth2",
"nautobot.core.authentication.ObjectPermissionBackend",
]
SOCIAL_AUTH_AZUREAD_TENANT_OAUTH2_KEY = os.getenv("SOCIAL_AUTH_AZUREAD_TENANT_OAUTH2_KEY", "")
SOCIAL_AUTH_AZUREAD_TENANT_OAUTH2_SECRET = os.getenv("SOCIAL_AUTH_AZUREAD_TENANT_OAUTH2_SECRET", "")
SOCIAL_AUTH_AZUREAD_TENANT_OAUTH2_TENANT_ID = os.getenv("SOCIAL_AUTH_AZUREAD_TENANT_OAUTH2_TENANT_ID", "")
```

<img width="362" alt="image" src="https://github.com/user-attachments/assets/e6a244fb-e19c-43dd-ac40-e724dcfa97b0">

**Comment by jdrew82** [comment]
@sabinayakc What about your group_sync function and the configuration for `SOCIAL_AUTH_PIPELINE`?

**Reply by sabinayakc** [reply]
When I add the logging section mentioned in the Readme. My application fails to initialize. 

![image](https://github.com/user-attachments/assets/956933e5-aaeb-4b4e-9a90-3c660cf17308)

Config looks like
```
SOCIAL_AUTH_PIPELINE = (
    "social_core.pipeline.social_auth.social_details",
    "social_core.pipeline.social_auth.social_uid",
    "social_core.pipeline.social_auth.auth_allowed",
    "social_core.pipeline.social_auth.social_user",
    "social_core.pipeline.user.get_username",
    "social_core.pipeline.user.create_user",
    "social_core.pipeline.social_auth.associate_user",
    "social_core.pipeline.social_auth.load_extra_data",
    "social_core.pipeline.user.user_details",
    "nautobot.extras.group_sync",
)
```

**Reply by sabinayakc** [reply]
Do I need to supply my own group_sync function? Or is there a default group sync function that is deployed.

When I put the above social auth pipeline configuration

I get
```
<class 'AttributeError'>

module 'nautobot.extras' has no attribute 'group_sync'

Python version: 3.11.10
Nautobot version: 2.3.11```

**Reply by glennmatthews** [reply]
The built-in `nautobot.extras.group_sync` module was just added recently in v2.3.12, which is why you're not seeing it in v2.3.11. Sorry for any confusion.

**Reply by sabinayakc** [reply]
Thank you for your response. I will upgrade my chart and let you know. 

https://github.com/nautobot/helm-charts/releases/tag/v2.4.1

**Reply by sabinayakc** [reply]
I'm getting this error. @glennmatthews 

![image](https://github.com/user-attachments/assets/aa0a5b22-aa5e-455b-8eac-a002db5c2fb8)

**Reply by glennmatthews** [reply]
Hm. I'll admit I haven't actually used that since it was added. :-) Try `"nautobot.extras.group_sync.group_sync"` instead?

**Reply by sabinayakc** [reply]
That worked.

We possibly need a PR to fix that in the core settings?

**Reply by glennmatthews** [reply]
Yes, perhaps it can be addressed as a part of #6623?

**Reply by sabinayakc** [reply]
@glennmatthews created a PR. https://github.com/nautobot/nautobot/pull/6624

**Comment by jdrew82** [comment]
@sabinayakc A few things:

- There appears to be a typo in the docs. That should be:

```
"my_custom_module": {"handlers": ["console", "file"], "level": "DEBUG"}
```

- You seem to misunderstand how the `group_sync()` function is intended to be added into the environment. It is a Python function that must be part of a Python module that you install into the environment. That is where the `my_custom_module` bit comes in for the logging. That is actually just an example name for whatever you wish to call the module you create, ie `my_group_sync_module` or whatever it is. Once you've created that and installed it into the environment, you need to update your SOCIAL_AUTH_PIPELINE to point to that module, ie it isn't in nautobot, it'd be `my_group_sync_module.group_sync.group_sync` assuming you're using the same structure as in the example. It basically needs to be able to be imported in the Python REPL on that path that you give it.

**Reply by sabinayakc** [reply]
As @glennmatthews  mentioned, I prefer not passing my own function and use the default one. Looks I need to use the latest helm chart that was deployed last week that uses v2.4.1

https://github.com/nautobot/helm-charts/releases/tag/v2.4.1

I will try that. Thank you for your prompt response.

**Reply by jdrew82** [reply]
I wasn't aware of the addition of group_sync into core Nautobot so I'd definitely go that route if you can be on 2.3.12.

**Reply by sabinayakc** [reply]
With this release, how would I enable logging? What would the handler and logger config look like? Like the module name.

**Reply by sabinayakc** [reply]
I found it. Thanks again.
`  "nautobot.extras.group_sync": {"handlers": ["normal_console"], "level": "DEBUG"}`
